testing/xproc3-house-rules.sch

```diff
@@ -123,15 +123,27 @@
    <!-- Any files to be reprieved from linking rules should be listed here, by /*/@name  -->
    <sch:let name="unlinked-xproc" value="('CONVERT-XML-REFERENCE-SET')"/>
    
+   <sch:let name="all-hrefs" value="//*[matches(@href, '^[^\}\{]+$')]/resolve-uri(@href, base-uri(.))"/>
+   
+   <!--<xsl:key name="reference-by-uri" match="*[matches(@href, '^[^\}\{]+$')]" use="resolve-uri(@href,base-uri(.))"/>-->
+   <!--<xsl:function name="ox:leads-with-variable-reference" as="xs:boolean">
+      <xsl:param name="v" as="item()"/>
+      <xsl:sequence select="string($v) => matches('^\s*\{\s*\$\i\c*\s*\}')"/>
+   </xsl:function>-->
+   
+   
    <sch:pattern>
       <!-- Not matching elements with href that contain { or } -->
       <sch:rule context="*[matches(@href, '^[^\}\{]+$')]">
          <sch:let name="exception" value="(/*/@name = $unlinked-xproc) or (tokenize(@href,'/')='lib')"/>
-         <sch:assert test="$exception or resolve-uri(@href, base-uri(.)) => unparsed-text-available()">No resource found at <sch:value-of
+         <sch:let name="expanded-uri" value="resolve-uri(@href, base-uri(.))"/>
+         <sch:assert test="$exception or ($expanded-uri => unparsed-text-available())">No resource found at <sch:value-of
             select="@href"/></sch:assert>
+         <sch:assert test="count($all-hrefs[. = $expanded-uri]) = 1">Repeat reference to this URI</sch:assert>
```

Comment:
Nitpick: is it better to say "Repeated" to avoid having the message sound like an imperative?

Is it not worth the trouble to have the message say where the repeated reference occurs, because it's not that hard for the user to find the duplication? Or maybe you just haven't gotten to that yet because it's lower priority.

Reply:
Finding the other/same reference(s) could be done but not using the present shortcut for testing I am using - and I already ripped out a key-based version (the one commented out) having found that SchXSLT under XProc seems to balk at it? So rather than debug that ...

... at present the assumption is that they will be close by or easily findable.

So far that is holding up and indeed the rule proves to be a handy one. 

Nevertheless I also switched it to warning level, as well. Since there is actually no formal rule against retrieving the same resource twice, and edge cases where one might wish to do so.

Reply:
Sounds good to me.